File: reorganize_json.py
import json
import os

path = r'c:\Users\Brian\Desktop\cliant\general.json'

with open(path, 'r', encoding='utf-8') as f:
    data = json.load(f)

std_trustees = 3
std_officials = 3

def process_group(group):
    # 1. Collect roles
    trustees = []
    officials = []
    members = []
    
    # Identify keys
    keys = list(group.keys())
    for key in keys:
        if key.startswith('trustee_'):
            trustees.append((key, group[key]))
        elif key.startswith('official_'):
            officials.append((key, group[key]))
        elif key.startswith('member_'):
            members.append((key, group[key]))
            
    # Sort them by their original indices if possible, or key name
    trustees.sort(key=lambda x: int(x[0].split('_')[1]))
    officials.sort(key=lambda x: int(x[0].split('_')[1]))
    members.sort(key=lambda x: int(x[0].split('_')[1]))
    
    # 2. Re-assign indices
    # Trustees always 1-3
    for i, (key, obj) in enumerate(trustees):
        obj['index'] = str(i + 1)
        
    # Officials always start at 4
    for i, (key, obj) in enumerate(officials):
        obj['index'] = str(std_trustees + i + 1)
        
    # Members always start at 7 (std_trustees + std_officials + 1)
    for i, (key, obj) in enumerate(members):
        obj['index'] = str(std_trustees + std_officials + i + 1)
        
    return group

# Recursively traverse the hierarchy
for county in data:
    for constituency in data[county]:
        for ward in data[county][constituency]:
            groups = data[county][constituency][ward]
            for group in groups:
                process_group(group)

with open(path, 'w', encoding='utf-8') as f:
    json.dump(data, f, indent=2)

print("Reorganization complete.")

File: reorganize_json.js
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'general.json');
const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));

const stdTrustees = 3;
const stdOfficials = 3;

function processGroup(group) {
    const keys = Object.keys(group);
    
    // Collect roles
    const trustees = keys.filter(k => k.startsWith('trustee_')).map(k => ({ oldKey: k, obj: group[k] })).sort((a,b) => parseInt(a.oldKey.split('_')[1]) - parseInt(b.oldKey.split('_')[1]));
    const officials = keys.filter(k => k.startsWith('official_')).map(k => ({ oldKey: k, obj: group[k] })).sort((a,b) => parseInt(a.oldKey.oldKey?.split('_')[1] || a.oldKey.split('_')[1]) - parseInt(b.oldKey.oldKey?.split('_')[1] || b.oldKey.split('_')[1]));
    const members = keys.filter(k => k.startsWith('member_')).map(k => ({ oldKey: k, obj: group[k] })).sort((a,b) => parseInt(a.oldKey.oldKey?.split('_')[1] || a.oldKey.split('_')[1]) - parseInt(b.oldKey.oldKey?.split('_')[1] || b.oldKey.split('_')[1]));

    // Remove old keys first
    trustees.forEach(t => delete group[t.oldKey]);
    officials.forEach(o => delete group[o.oldKey]);
    members.forEach(m => delete group[m.oldKey]);

    // Re-add with new keys and matching indices
    trustees.forEach((t, i) => {
        const index = (i+1).toString();
        const newKey = `trustee_${index}`;
        delete t.obj.index; // Remove redundant index numbering from data
        group[newKey] = t.obj;
    });

    officials.forEach((o, i) => {
        const index = (stdTrustees + i + 1).toString();
        const newKey = `official_${index}`;
        delete o.obj.index; // Remove redundant index numbering from data
        group[newKey] = o.obj;
    });

    members.forEach((m, i) => {
        const index = (stdTrustees + stdOfficials + i + 1).toString();
        const newKey = `member_${index}`;
        delete m.obj.index; // Remove redundant index numbering from data
        group[newKey] = m.obj;
    });
}

for (const county in data) {
    for (const constituency in data[county]) {
        for (const ward in data[county][constituency]) {
            data[county][constituency][ward].forEach(processGroup);
        }
    }
}

fs.writeFileSync(filePath, JSON.stringify(data, null, 2));
console.log('Key reorganization complete.');
